core/src/main/java/net/opentsdb/query/processor/downsample/DownsampleNumericToNumericArrayIterator.java

```diff
@@ -0,0 +1,577 @@
+ * always be the midnight hour at the start of the day and includes values from
+ * [midnight of day, midnight of next day). This implies:
+ * <ul>
+ * <li>If a source timestamp is earlier than the {@link DownsampleConfig#start()}
+ * it will not be included in the results even though the query may have a start
+ * timestamp earlier than {@link DownsampleConfig#start()} (due to the fact that
+ * the config will snap to the earliest interval greater than or equal to the
+ * query start timestamp.</li>
+ * <li>If a source timestamp is later than the {@link DownsampleConfig#end()}
+ * time but is within the interval defined by {@link DownsampleConfig#end()},
+ * it <b>will</b> be included in the results.</li>
+ * </ul></li>
+ * </ul>
+ * <p>
+ * @since 3.0
+ */
+public class DownsampleNumericToNumericArrayIterator implements QueryIterator, TimeSeriesValue<NumericArrayType> {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(DownsampleNumericToNumericArrayIterator.class);
+  /** The result we belong to. */
+  private final DownsampleResult result;
+
+  /** The downsampler config. */
+  private final DownsampleConfig config;
+
+  /** Whether or not the iterator has another real or filled value. */
+  private boolean has_next;
+
+  /** The current interval start timestamp. */
+  private TimeStamp interval_start;
+
+  /** The current interval end timestamp. */
+  private TimeStamp interval_end;
+
+  /** The iterator pulled from the source. */
+  private TypedTimeSeriesIterator<? extends TimeSeriesDataType> iterator;
+
+  /** The partial aggregation defaults for double values */
+  private static final double[] partial_doubles =
+      {0.0, 0.0, Double.MAX_VALUE, Double.MIN_VALUE, 0.0, -1.0};
+
+  /** The partial aggregation defaults for long values */
+  private static final long[] partial_longs = {0, 0, Long.MAX_VALUE, Long.MIN_VALUE, 0, -1};
+
+  /** The thread local accumulator to hold the raw data points for partial aggregations */
+  private static final ThreadLocal<Accumulator> threadLocalAccs =
+      ThreadLocal.withInitial(() -> new Accumulator());
+
+  /** The DS to contain partial aggregates based on the aggregation function for double values */
+  private static final ThreadLocal<double[]> threadLocalDoubleAggs =
+      ThreadLocal.withInitial(() -> Arrays.copyOf(partial_doubles, partial_doubles.length));
+
+  /** The DS to contain partial aggregates based on the aggregation function for long values */
+  private static final ThreadLocal<long[]> threadLocalLongAggs =
+      ThreadLocal.withInitial(() -> Arrays.copyOf(partial_longs, partial_longs.length));
+
+  /** The downsample query node */
+  private QueryNode node;
+
+  /** Number of downsample intervals */
+  protected int intervals;
+
+  /**
+   * Enums for aggregation functions
+   *
+   */
+  protected static enum AggEnum {
+    sum, count, min, max, last, avg;
+  }
+
+  /** Enum identifying the aggregation function */
+  private final AggEnum aggEnum;
```

Comment:
We should keep this out and generic. Some queries will be coming in with percentile aggs (even though I haven't implemented those as arrays yet) and more.